### ISSUE TITLE
--enable-pretty-printing breaks gdb for complex objects used as keys in maps

### ISSUE THREAD

**Issue by gjblanco** [microsoft/vscode-cpptools#3024]
**OS:**
 Ubuntu 18.04

**VS Code:**
Version: 1.30.2
Commit: 61122f88f0bf01e2ac16bdb9e1bc4571755f5bd8
Date: 2019-01-07T22:49:48.319Z
Electron: 2.0.12
Chrome: 61.0.3163.100
Node.js: 8.9.3
V8: 6.1.534.41
OS: Linux x64 4.15.0-43-generic

Extensions: 
ms-vscode.cpptools

When in debug mode, after clicking the |> (little triangle) next to a variable of type std::map<MyObject, int>, where MyObject is a class that has a custom comparator, the message "Stopping due to fatal error: AggregateException: One or more errors occurred." shows up in the debug terminal, and it just hangs there until you stop it.

It works fine with gdb (i.e if you gdb main and run until after the variable is declared, "print variable" works)

It also works fine if you set "setupCommands": [] in launch.json (i.e disable pretty printing)

**To Reproduce**

1. Use the files below: copy main.cpp and makefile into a folder named "small" (or change the launch.json file below according to your file structure)
2. copy the "launch.json" given below to your own launch.json
3. cd into small and make
4. put a breakpoint on the second m[...] = ... (in the file below the line is m[{3, 4}] = 0; )
5. start debugging. get to the breakpoint.
6. click on the triangle in the local variables area next to m. it should hang. An error should appear in the debug terminal.

```
// launch.json:
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [

        {
            "name": "(gdb) Small",
            "type": "cppdbg",
            "request": "launch",
            "program": "${workspaceFolder}/small/main",
            "args": [],
            "stopAtEntry": false,
            "cwd": "${workspaceFolder}",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
                
            ]
        }
    ]
}
```

```
//main.cpp
#include <map>

using std::map;

struct X {
    int a, b;
};

struct Xcmp {
    bool operator()(const X& x1, const X& x2) {
        return x1.a < x2.a || (x1.a == x2.a && x1.b < x2.b);
    }
};

int main() {
    map<X, int, Xcmp> m;
    m[{1, 2}] = 0;
    m[{3, 4}] = 0;
}
```

```
# makefile:
all: main

main: main.cpp
	g++ -g main.cpp -o main

clean:
	rm -rf main
```

**Comment by gjblanco** [comment]
I just checked, and it doesn't work with the latest insiders version

**Comment by pieandcakes** [comment]
The problem is in line https://github.com/Microsoft/MIEngine/blob/29c6569861b2042c43f04d8d78b2566db2e2ef1f/src/MIDebugEngine/Engine.Impl/Variables.cs#L606 . 

We end up in this code path because in `pretty-printing` mode we end up with the `displayhint="map"` being sent by `gdb` which causes us to process it in that line where if you disable `pretty-printing`, then the map processing is not done.

Once in the map processing portion, we mistakenly assume that the first child, the key, is a simple type. 

I filed https://github.com/Microsoft/MIEngine/issues/822 and will use that to track this issue.

**Comment by sean-mcmanus** [comment]
This debugger bug should be fixed with 0.22.0-insiders. You can get it via setting the C_Cpp.updateChannel setting to "Insiders", or downloading it manually from https://github.com/Microsoft/vscode-cpptools/releases .
